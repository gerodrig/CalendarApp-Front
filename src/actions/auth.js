import { fetchNoToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";
import Swal from "sweetalert2";
import { eventLogout } from "./event";

export const startLogin = (email, password) => {
	return async (dispatch) => {
		const response = await fetchNoToken(
			"auth",
			{ email, password },
			"POST"
		);

		const body = await response.json();

		if (body.ok) {
			localStorage.setItem("token", body.token);
			localStorage.setItem("token-init-date", new Date().getTime());

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				})
			);
		} else {
			Swal.fire({
				icon: "error",
				title: "Error",
				text: body.message,
				confirmButtonColor: "#0062cc",
			});
		}
	};
};

export const startRegister = (name, email, password, password2) => {
	return async (dispatch) => {
		if (password !== password2) {
			return Swal.fire({
				icon: "error",
				title: "Error",
				text: "Passwords do not match",
				confirmButtonColor: "#0062cc",
			});
		}

		const response = await fetchNoToken(
			"auth/new",
			{ name, email, password },
			"POST"
		);

		const body = await response.json();

		if (body.ok) {
			localStorage.setItem("token", body.token);
			localStorage.setItem("token-init-date", new Date().getTime());

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				})
			);
		} else {
			Swal.fire({
				icon: "error",
				title: "Error",
				text: body.message,
				confirmButtonColor: "#0062cc",
			});
		}
	};
};

export const startChecking = () => {
	return async (dispatch) => {

		const response = await fetchWithToken("auth/renew");
		const body = await response.json();

		if (body.ok) {

			localStorage.setItem("token", body.token);
			localStorage.setItem("token-init-date", new Date().getTime());

			dispatch(
				login({
					uid: body.uid,
					name: body.name,
				})
			);
		} else {
			dispatch(checkingFinish());
		}
	};
};

export const startLogout = () => {
	return async (dispatch) => {
		
		localStorage.clear();
		dispatch(logoutUser());
		dispatch( eventLogout() );

	};
};

const checkingFinish = () => ({ type: types.authCheckingFinish });


const login = (user) => ({
	type: types.authLogin,
	payload: user,
});

const logoutUser = () => ({
	type: types.authLogout,
});

