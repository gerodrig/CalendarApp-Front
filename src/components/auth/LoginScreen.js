import React from "react";
import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import { startLogin } from "../../actions/auth";
import { useForm } from "../../hooks/useForm";
import "./login.css";

export const LoginScreen = () => {
	const dispatch = useDispatch();

	const [formLoginValues, handleLoginInputChange] = useForm({
		loginEmail: "",
		loginPassword: "",
	});

	const { loginEmail, loginPassword } = formLoginValues;	

	const handleLogin = (e) => {
		e.preventDefault();

		dispatch(startLogin(loginEmail, loginPassword));
	};

	return (
		<div className="container login-container">
			<div className="row">
				<div className="col-md-6 login-form-1">
					<h3>Login</h3>
					<form onSubmit={handleLogin}>
						<div className="form-group">
							<input
								type="text"
								className="form-control"
								placeholder="Email"
								name="loginEmail"
								value={loginEmail}
								onChange={handleLoginInputChange}
							/>
						</div>
						<div className="form-group mt-3">
							<input
								type="password"
								className="form-control"
								placeholder="Password"
								name="loginPassword"
								value={loginPassword}
								onChange={handleLoginInputChange}
							/>
						</div>
						<div className="form-group mt-3 d-flex justify-content-center">
							<input
								type="submit"
								className="btnSubmit"
								value="Login"
							/>
						</div>
						<Link
							to="/register"
							className="link mt-3 d-flex justify-content-center"
						>
							Create new account
						</Link>
					</form>
				</div>
			</div>
		</div>
	);
};
