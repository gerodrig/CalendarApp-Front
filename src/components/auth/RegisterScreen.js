import React from 'react'
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom'
import { startRegister } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const RegisterScreen = () => {

    const dispatch = useDispatch();

    const [formRegisterValues, handleRegisterInputChange] = useForm({
		registerName: "Gerry",
		registerEmail: "gerry@outlook.com",
		registerPassword: "",
		registerPassword2: "",
	});

	const { registerName, registerEmail, registerPassword, registerPassword2 } = formRegisterValues;

    const handleRegister = (e) => {
        e.preventDefault();

        

        dispatch(startRegister(registerName, registerEmail, registerPassword, registerPassword2));
    };


    return (
        <div className="container login-container">
            <div className="row">
                <div className="col-md-6 login-form-2">
                    <h3>Register</h3>
                    <form onSubmit={ handleRegister }>
                        <div className="form-group">
                            <input
                                type="text"
                                className="form-control"
                                placeholder="Name"
                                name="registerName"
                                value={registerName}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="email"
                                className="form-control mt-3"
                                placeholder="Email"
                                name="registerEmail"
                                value={registerEmail}
                                onChange={handleRegisterInputChange}
                            />
                        </div>
                        <div className="form-group">
                            <input
                                type="password"
                                className="form-control mt-3"
                                placeholder="Password"
                                name="registerPassword"
                                value={registerPassword}
                                onChange={handleRegisterInputChange}                                
                            />
                        </div>

                        <div className="form-group mt-3">
                            <input
                                type="password"
                                className="form-control"
                                placeholder="Re-enter password" 
                                name="registerPassword2"
                                value={registerPassword2}
                                onChange={handleRegisterInputChange}
                            />
                        </div>

                        <div className="form-group mt-3 d-flex justify-content-center">
                            <input 
                                type="submit" 
                                className="btnSubmit" 
                                value="Create account" />
                        </div>
                        <Link   to="/login"
                                className="link mt-3 d-flex justify-content-center"
                                style={ { color: '#fff' } }>
                        Login
                        </Link>
                    </form>
                </div>
            </div>
        </div>
    )
}
