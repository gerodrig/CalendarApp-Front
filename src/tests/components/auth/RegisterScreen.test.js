import React from "react";
import Swal from "sweetalert2";
import { mount } from "enzyme";
import { Provider } from "react-redux";

import "@testing-library/jest-dom";

import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { RegisterScreen } from "../../../components/auth/RegisterScreen";
import { BrowserRouter as Router } from "react-router-dom";
import { startRegister } from "../../../actions/auth";

jest.mock("../../../actions/auth", () => ({
	startRegister: jest.fn(),
}));

jest.mock("sweetalert2", () => ({
	fire: jest.fn(),
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initialState = {};
const store = mockStore(initialState);
store.dispatch = jest.fn();

const wrapper = mount(
	<Provider store={store}>
		<Router>
			<RegisterScreen />
		</Router>
	</Provider>
);

describe("Tests in <LoginScreen /> component", () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	test("should render component correctly", () => {
		expect(wrapper).toMatchSnapshot();
	});

	test("should not register if passwords are different", () => {
		//1-2 .. password1 and password2 simutlates the same input
		wrapper.find('input[name="registerPassword"]').simulate("change", {
			target: {
				name: "registerPassword",
				value: "123456",
			},
		});
		wrapper.find('input[name="registerPassword2"]').simulate("change", {
			target: {
				name: "registerPassword2",
				value: "1234567",
			},
		});

		wrapper.find("form").at(0).prop("onSubmit")({ preventDefault() {} });

		//startRegister is not called
		expect(startRegister).toHaveBeenCalled();

		//Swal fire is called from the backend
	});

	test("should register if passwords are the same", () => {
		wrapper.find('input[name="registerPassword"]').simulate("change", {
			target: {
				name: "registerPassword",
				value: "123456",
			},
		});
		wrapper.find('input[name="registerPassword2"]').simulate("change", {
			target: {
				name: "registerPassword2",
				value: "123456",
			},
		});

		wrapper.find("form").at(0).prop("onSubmit")({ preventDefault() {} });

		//startRegister is not called
		expect(Swal.fire).not.toHaveBeenCalled();
		expect(startRegister).toHaveBeenCalled();
	});
});
