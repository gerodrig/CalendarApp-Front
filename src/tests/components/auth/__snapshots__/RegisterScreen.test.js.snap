// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <LoginScreen /> component should render component correctly 1`] = `
<div
  className="container login-container"
>
  <div
    className="row"
  >
    <div
      className="col-md-6 login-form-2"
    >
      <h3>
        Register
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="registerName"
            onChange={[Function]}
            placeholder="Name"
            type="text"
            value="Gerry"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control mt-3"
            name="registerEmail"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value="gerry@outlook.com"
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control mt-3"
            name="registerPassword"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group mt-3"
        >
          <input
            className="form-control"
            name="registerPassword2"
            onChange={[Function]}
            placeholder="Re-enter password"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group mt-3 d-flex justify-content-center"
        >
          <input
            className="btnSubmit"
            type="submit"
            value="Create account"
          />
        </div>
        <a
          className="link mt-3 d-flex justify-content-center"
          href="/login"
          onClick={[Function]}
          style={
            Object {
              "color": "#fff",
            }
          }
        >
          Login
        </a>
      </form>
    </div>
  </div>
</div>
`;
